core/endorser/endorser.go

```diff
@@ -472,86 +473,124 @@ func (e *Endorser) ProcessProposalSuccessfullyOrError(up *UnpackedProposal) (*pb
+				if len(keyPolicies) > 0 {
+					// For simplicity, we'll always add the SBE policies to the public ChaincodeCall, and set the disregard flag if the chaincode policy is not required.
+					ccCall.KeyPolicies = keyPolicies
+					if !policies.requireChaincodePolicy(chaincode) {
+						ccCall.DisregardNamespacePolicy = true
+					}
+				}
+			} else {
 				// Since each collection in a chaincode could have different values of the NoPrivateReads flag, create a new Chaincode entry for each.
-				ccInterest.Chaincodes = append(ccInterest.Chaincodes, &pb.ChaincodeCall{
-					Name:            chaincode,
-					CollectionNames: []string{collectionName},
-					NoPrivateReads:  !isRead,
-				})
+				ccCall.CollectionNames = []string{collection}
+				ccCall.NoPrivateReads = !simResult.PrivateReads.Exists(chaincode, collection)
 			}
+			ccInterest.Chaincodes = append(ccInterest.Chaincodes, ccCall)
 		}
 	}
 
-	// add the key signature policies - at the moment they are grouped together in the first chaincode
-	// TODO implement DisregardNamespacePolicy hint in ChaincodeCall
-	if len(ccInterest.Chaincodes) > 0 {
-		for _, policyBytes := range simResult.KeySignaturePolicies {
-			policy, err := protoutil.UnmarshalSignaturePolicy(policyBytes)
-			if err != nil {
-				return nil, err
+	endorserLogger.Info("ccInterest", ccInterest)
+	return ccInterest, nil
+}
+
+type metadataPolicies struct {
+	// map of SBE policies: namespace -> array of policies
+	sbePolicies map[string][]*common.SignaturePolicyEnvelope
+	// whether the chaincode/collection policy is required for endorsement: namespace -> collection -> isRequired
+	// empty collection name represents the public rwset
+	policyRequired map[string]map[string]bool
+}
+
+func parseWritesetMetadata(metadata ledger.WritesetMetadata) (*metadataPolicies, error) {
+	mp := &metadataPolicies{
+		sbePolicies:    map[string][]*common.SignaturePolicyEnvelope{},
+		policyRequired: map[string]map[string]bool{},
+	}
+	for ns, cmap := range metadata {
+		mp.policyRequired[ns] = map[string]bool{}
+		sbePolicies := []*common.SignaturePolicyEnvelope{}
+		for coll, kmap := range cmap {
+			// look through each of the states that were written to
+			for _, stateMetadata := range kmap {
+				// if the state metadata is nil, or it doesn't contain data relating to SBE policy, then this chaincode/collection policy is required
+				if stateMetadata == nil {
+					mp.policyRequired[ns][coll] = true
+					continue
+				}
+				if policyBytes, sbeExists := stateMetadata[pb.MetaDataKeys_VALIDATION_PARAMETER.String()]; sbeExists {
+					policy, err := protoutil.UnmarshalSignaturePolicy(policyBytes)
+					if err != nil {
+						return nil, err
+					}
+					sbePolicies = append(sbePolicies, policy)
+					if !mp.policyRequired[ns][""] {
+						// the SBE policies need a container in the proto message - false means namespace policy will be disregarded (unless later set to true)
+						mp.policyRequired[ns][""] = false
+					}
```

Comment:
You're right, it doesn't need to be in the inner loop.  But it can't be the default because a 'false' entry here is only required if there are SBE policies present.  I've moved this to the end and put it in a conditional.  I've added more comments to explain the data structure and logic a bit better.

Reply:
Fixed in the latest patch.